src/org/sparkboard/server/server.clj

```diff
@@ -237,12 +240,28 @@
       (.doFinal (.getBytes message "UTF-8"))
       (Hex/encodeHexString)))
 
+(defn wrap-saved-body
+  "Duplicates body of urlencoded requests for later parsing.
+   Must go before/outside ring middleware (i.e. after in the `->`
+  thread), so it still has access to untouched `body` of request."
+  [f]
+  (fn [req]
+    (log/warn "[wrap-saved-body" (ring.util.request/urlencoded-form? req))
+    (if (ring.util.request/urlencoded-form? req)
+      (f (let [input (slurp (:body req)
+                            :encoding (or (ring.util.request/character-encoding req)
+                                          "UTF-8"))]
+           (assoc req
+                  :body  (io/input-stream (.getBytes input))
+                  :body2 (io/input-stream (.getBytes input)))))
```

Comment:
how about `:body-string` here, with `input`? saves creating an extra stream that can also only be read once

Reply:
good point, I actually intended to make that refactor earlier in the process